Fix issues #392 and #400 caused by cleanfigure

PeterPablo opened the pull request:
Existing ACID(21) `pruneOutsideBox()` is fixed and ACID(7) `linesWithOutliers` keeps working:
![image](https://cloud.githubusercontent.com/assets/4340267/6379445/a040d85c-bd32-11e4-8669-a9919817748a.png)
![image](https://cloud.githubusercontent.com/assets/4340267/6379478/ee80e0b6-bd32-11e4-8216-777a25737483.png)

Fixes issues #392 and #400.

**edit:** Thanks go to @akloeckner for providing the missing real fix (PR #547). I incorporated it here, since I added quite some comments and restructured the code a little bit.

nschloe commented:
Is there a reason why the revert is part of this?

PeterPablo commented:
No there is not :-( My git magic is struggling with  me. Only the three commits 494ff4a, f1a0df6 and ad00575 should go into this PR.

PeterPablo commented:
Ok, I think I got it sorted out with a dangerous `git push --force`... The pull request now contains only the three relevant commits.

okomarov commented:
I think the additional ACID is not strictly necessary since we can edit ACID(21) to improve the explanation, or am I missing something?

Also, the description in the new ACID is not very informative and maybe this is why I do not get in what it is different from ACID(21).

PeterPablo commented:
The  new test is indeed not strictly necessary but helped me track down the issue. A new aspect is the red plot that has data points inside the axis limits. The crucial comments in the new test are `% some points are outside visible axis` and `% all points are outside visible axis`.
This is not covered by ACID(21). Certainly ACID(21) could be adapted, but I didn't want to go through the hassle.

okomarov commented:
ACID(21) has the blue line with points inside the axis. 

By description I meant `stats.description`.

PeterPablo commented:
Please wait with merging. I just discovered that ACID(7) fails. This is probably the reason for the check I removed in d42ac8447f5bb04ee82fc2e45eddc85921e5c9df.

PeterPablo commented:
@okomarov You are right. Totally forgot about that blue sine :-) I guess we can drop the new test then.

PeterPablo commented:
I had another look and the issue  is tougher than it seemed at first. The new test does cover a new aspect, namely a line of two outlying points that intersect the rectangle defined by the axis.
As the last commit indicates code for the calculation of the boolean needs to be added to actually fix the problem. Otherwise either ACID(7) or the newly added ACID(100) fails.
Pointer for implementation idea: http://blogs.mathworks.com/loren/2011/08/29/intersecting-lines/

PeterPablo commented:
Now I am certain that the newly added test can be removed again since both aspects of moving points closer to the axis (`movePointsCloser()`) are covered by ACID(7) and ACID(22). I spent more time than intended on figuring out the base of this and am unsure if I can finish this up timely. That's why I pushed that preliminary commit to write down the findings and write down a sketch of a viable approach.

okomarov commented:
As you said, the problem is more complicated than it seems. I will try to think about the most common cases we need to tackle. We also need to consider that people do not necessarily use `cleanfigure()` and thus try to give priority in the algorithm to some realistic scenario, leaving corner cases to potential workarounds in the workflow, rather than trying to solve everything.

PeterPablo commented:
As far as I can see I reached the bottom of the issue and documented it in the code with a boolean flag that makes one of the two named ACID tests fail. I described there how to derive the state of that boolean. It isn't overly complicated, but still one has to write the code...

PeterPablo commented:
I rebased this PR, removed the newly added testcase, applied the fix from @akloeckner and updated the initial post to reflect this. Could please somebody else check my changes?
